### PR TITLE
Fuzzer: Add an option to fuzz with initial wasm contents

### PR DESCRIPTION
**Pull request by kripken**
Previously the fuzzer constructed a new random valid wasm file from
scratch. The new `--initial-fuzz=FILENAME` option makes it start from
an existing wasm file, and then add random contents on top of that. It
also randomly modifies the existing contents, for example tweaking
a Const, replacing some nodes with other things of the same type, etc.
It also has a chance to replace a drop with a logging (as some of our
tests just drop a result, and we match the optimized output's wasm
instead of the result; by logging, the fuzzer can check things).

The goal is to find bugs by using existing hand-written testcases as
a basis. This PR uses the test suite's testcases as initial fuzz contents.
This can find issues as they often check for corner cases - they are
designed to be "interesting", which random data may be less likely
to find.

This has found several bugs already, see recent fuzz fixes. I mentioned
the first few on Twitter but past 4 I stopped counting...

https://twitter.com/kripken/status/1314323318036602880

This required various changes to the fuzzer's generation to account
for the fact that there can be existing functions and so forth before
it starts to run, so it needs to avoid collisions and so forth.

### CODE REVIEW

File: src/tools/fuzzing.h [hunk 1]
```diff
@@ -507,11 +607,28 @@ class TranslateToFuzzReader {
...
+    labelIndex = 0;
+    assert(breakableStack.empty());
+    assert(hangStack.empty());
+  }
+
+  void finishCreatingFunctionContents() {
+    if (HANG_LIMIT > 0) {
+      addHangLimitChecks(func);
+    }
+    typeLocals.clear();
```
Comment (tlively) [inline src/tools/fuzzing.h:621]: Clearing function-local state manually seems like a code smell. If we need to add more state in the future, it would be easy to forget to clear it. Could this be handled automatically via RAII with some sort of `FunctionCreationContext` object?

File: scripts/fuzz_opt.py [hunk 2]
```diff
@@ -115,10 +118,18 @@ def randomize_feature_opts():
     print('randomized feature opts:', ' '.join(FEATURE_OPTS))
 
 
+# the optimizations to run on the wasm
 FUZZ_OPTS = None
+# whether NaN values are allowed, or we de-NaN them
```
[the remark this answers is not in the document]
Comment (kripken) [reply scripts/fuzz_opt.py:123]: I agree this isn't great. But `None` is meant to indicate a null value. It will always be replaced later. So I worry that setting to `False` or `True`, while technically valid, could be confusing as it could indicate a default, but it isn't that.

Another option is to not declare these in the global scope at all, and move the comments to where these are assigned. I'll push a commit with that, let me know what you think.

File: src/tools/fuzzing.h [hunk 3]
```diff
@@ -406,47 +414,132 @@ class TranslateToFuzzReader {
...
       auto* global =
         builder.makeGlobal(std::string("global$") + std::to_string(index),
                            type,
                            makeConst(type),
                            Builder::Mutable);
-      wasm.addGlobal(global);
-      globalsByType[type].push_back(global->name);
+      if (!wasm.getGlobalOrNull(global->name)) {
+        wasm.addGlobal(global);
+        globalsByType[type].push_back(global->name);
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/tools/fuzzing.h:458]: This code adds some random globals, so it seemed ok to just skip any name collisions. But I guess now we have the name generator, I should use that - will push a commit in a bit.

File: src/tools/fuzzing.h (hunk 1 above)
Comment (kripken) [reply src/tools/fuzzing.h:621]: Fair enough, but how about in a later PR? Most of this is pre-existing, and this PR is already doing a lot.

File: src/tools/fuzzing.h [hunk 4]
```diff
@@ -406,47 +414,132 @@ class TranslateToFuzzReader {
...
+      // maximum larger than the initial.
+      // TODO: scan the wasm for grow instructions?
+      wasm.memory.max =
+        std::min(Address(wasm.memory.initial + 1), Address(Memory::kMaxSize32));
+    }
+    // Avoid an unlimited memory size, which would make fuzzing very difficult
+    // as different VMs will run out of system memory in different ways.
+    if (wasm.memory.max == Memory::kUnlimitedSize) {
+      wasm.memory.max = wasm.memory.initial;
+    }
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/tools/fuzzing.h:510]: Good point, fixing!

File: scripts/fuzz_opt.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (tlively) [reply scripts/fuzz_opt.py]: Oh, I see. It might be good to reword the comments to be more explicit about what values will eventually be assigned to these globals, since right now they sound like plain Boolean options.

File: src/tools/fuzzing.h (hunk 1 above)
Comment (tlively) [reply src/tools/fuzzing.h:619]: Sure, sounds good.

File: scripts/fuzz_opt.py [hunk 5]
```diff
@@ -402,6 +495,11 @@ def can_run(self, wasm):
                 # relatively slow, so run it less frequently
                 if random.random() < 0.5:
                     return False
+                # wasm2c name mangling must be fixed for us to support initial
+                # content, which may include testcases with names that need
+                # mangling. https://github.com/WebAssembly/binaryen/pull/3228
+                if INITIAL_CONTENTS:
+                    return False
```
Comment (aheejin) [inline scripts/fuzz_opt.py:502]:
```suggestion
```
I guess we can remove this after we land #3228?

File: src/tools/fuzzing.h [hunk 6]
```diff
@@ -432,21 +461,83 @@ class TranslateToFuzzReader {
...
+      wasm.memory.max = wasm.memory.initial;
     }
+    if (wasm.memory.max <= wasm.memory.initial) {
+      // To allow growth to work (which a testcase may assume), try to make the
+      // maximum larger than the initial.
+      // TODO: scan the wasm for grow instructions?
+      wasm.memory.max =
+        std::min(Address(wasm.memory.initial + 1), Address(Memory::kMaxSize32));
+    }
+    // See above for globals.
```
Comment (aheejin) [inline src/tools/fuzzing.h:509]: Does this mean we don't allow imported memory? It this is all I guess we can just say that here, because it's one-line anyway.. The same for the same comment in `finalizeTable`.

**Review by aheejin — COMMENTED** [review]
Haven't gotten to the end yet, but some comments and questions

File: scripts/fuzz_opt.py [hunk 7]
```diff
@@ -159,6 +169,80 @@ def randomize_fuzz_settings():
...
+        'simd.wast',
+        # corner cases of escaping of names is not interesting
+        'names.wast',
+        # huge amount of locals that make it extremely slow
+        'too_much_for_liveness.wasm',
+        # these contain illegal pops()
+        # https://github.com/WebAssembly/binaryen/issues/3213
+        'instrument-locals_all-features.wast',
+        'remove-unused-names_code-folding_all-features.wast',
+        'Os_print-stack-ir_all-features.wast'
```
Comment (aheejin) [inline scripts/fuzz_opt.py:201]:
```suggestion
        'too_much_for_liveness.wasm'
```
After #3282 lands there will be no invalid pops.

File: src/tools/fuzzing.h [hunk 8]
```diff
@@ -783,10 +898,81 @@ class TranslateToFuzzReader {
     ReFinalize().walkFunctionInModule(func, &wasm);
   }
 
+  void modifyInitialFunctions() {
+    if (wasm.functions.empty()) {
+      return;
+    }
+    // Pick a chance to fuzz the contents of a function.
+    const int RESOLUTION = 10;
+    auto chance = upTo(RESOLUTION + 1);
```
Comment (aheejin) [inline src/tools/fuzzing.h:907]: Is `chance` the same across all functions? (Not that it's a problem, but just curious if you intended it to be within the loop or not)

File: src/tools/fuzzing.h [hunk 9]
```diff
@@ -783,10 +898,81 @@ class TranslateToFuzzReader {
...
+        // We can't allow extra imports, as the fuzzing infrastructure wouldn't
+        // know what to provide.
+        func->module = func->base = Name();
+        func->body = make(func->sig.results);
+      }
+      // Optionally, fuzz the function contents.
+      if (upTo(RESOLUTION) >= chance) {
+        dropToLog(func);
+        // TODO add some locals? and the rest of addFunction's operations?
+        // TODO: interposition, replace initial a(b) with a(RANDOM_THING(b))
```
Comment (aheejin) [inline src/tools/fuzzing.h:921]: What is `a(b)` notation? Does this mean `a` and `b` are functions and `b` comes after `a`..?

File: src/tools/fuzzing.h [hunk 10]
```diff
@@ -783,10 +898,81 @@ class TranslateToFuzzReader {
...
+        recombine(func);
+        mutate(func);
+        fixLabels(func);
```
Comment (aheejin) [inline src/tools/fuzzing.h:924]: It [seems](https://github.com/WebAssembly/binaryen/blob/2dc156d1d3af7f3eb40bdce7c20aa84c98d50c15/src/tools/fuzzing.h#L548-L559) we do these transformations only when we allow OOB. Is it OK not to check that here too?

File: scripts/fuzz_opt.py [hunk 11]
```diff
@@ -159,6 +169,80 @@ def randomize_fuzz_settings():
...
+            chosen = random.choice(split_parts)
+            module, asserts = chosen
+            if not module:
+                # there is no module in this choice (just asserts), ignore it
+                print('initial contents has no module')
+                return
+            test_name = 'initial.wat'
+            with open(test_name, 'w') as f:
+                f.write(module)
+            print('  picked submodule of wast, of size', len(module))
```
[the remark this answers is not in the document]
Comment (kripken) [reply scripts/fuzz_opt.py]: Good point, it's odd to do it here and not there.

Thinking on this, how about if this prints the index chosen, and not the size? That seems more useful anyhow, and less ambiguous.

File: scripts/fuzz_opt.py [hunk 12]
```diff
@@ -361,23 +432,45 @@ class CompareVMs(TestCaseHandler):
...
 
-        class Wasm2C(VM):
+            def run(self, wasm):
+                run([in_bin('wasm-opt'), wasm, '--emit-js-wrapper=' + wasm + '.js'] + FEATURE_OPTS)
+                return run_vm([shared.V8, wasm + '.js'] + shared.V8_OPTS + ['--', wasm])
+
+            def can_run(self, wasm):
+                # INITIAL_CONTENT is disallowed because some initial spec testcases
+                # have names that require mangling, see
+                # https://github.com/WebAssembly/binaryen/pull/3216
```
[the remark this answers is not in the document]
Comment (kripken) [reply scripts/fuzz_opt.py:456]: The specific problem is that an export with "-" in it is not a valid JS name. It can be used with quotes, but then there are other issues, like not working in an ES6 module import, it turns out (see link).

File: scripts/fuzz_opt.py [hunk 13]
```diff
@@ -638,6 +729,13 @@ def run(self, wasm):
         return run_vm([shared.NODEJS, js_file, 'a.wasm'])
 
     def can_run_on_feature_opts(self, feature_opts):
+        # TODO: properly handle memory growth. right now the wasm2js handler
+        # uses --emscripten which assumes the Memory is created before, and
+        # wasm2js.js just starts with a size of 1 and no limit. We should switch
+        # to non-emscripten mode or adding memory information, or check
+        # specifically for growth here
+        if INITIAL_CONTENTS:
+            return False
```
[the remark this answers is not in the document]
Comment (kripken) [reply scripts/fuzz_opt.py:729]: Not directly in itself, but initial testcases can use growth, if the wasm in the initial testcase uses it.

File: src/tools/fuzzing.h [hunk 14]
```diff
@@ -406,21 +414,42 @@ class TranslateToFuzzReader {
...
   void setupTable() {
     wasm.table.exists = true;
+    wasm.table.initial = wasm.table.max = 0;
     wasm.table.segments.emplace_back(builder.makeConst(int32_t(0)));
   }
 
   std::map<Type, std::vector<Name>> globalsByType;
 
   void setupGlobals() {
+    // If there were initial wasm contents, there may be imported globals.
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/tools/fuzzing.h]: I'll improve the comment. The issue is that we'll fail to run the wasm if it needs an import we don't provide. It's simplest to remove any imports the fuzz harness doesn't provide by default.

File: src/tools/fuzzing.h [hunk 15]
```diff
@@ -432,21 +461,83 @@ class TranslateToFuzzReader {
...
-      wasm.addEvent(event);
+      auto name = std::string("event$") + std::to_string(i);
+      if (!wasm.getEventOrNull(name)) {
+        auto* event =
+          builder.makeEvent(name,
+                            WASM_EVENT_ATTRIBUTE_EXCEPTION,
+                            Signature(getControlFlowType(), Type::none));
+        wasm.addEvent(event);
+      }
+    }
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/tools/fuzzing.h:474]: Sounds good, done.

File: src/tools/fuzzing.h [hunk 16]
```diff
@@ -432,21 +461,83 @@ class TranslateToFuzzReader {
...
+  }
+
+  void finalizeMemory() {
+    for (auto& segment : wasm.memory.segments) {
+      Address maxOffset = segment.data.size();
+      if (!segment.isPassive) {
+        // If the offset is a global that was imported (which is ok) but no
+        // longer is (not ok) we need to change that.
+        if (auto* offset = segment.offset->dynCast<GlobalGet>()) {
+          if (!wasm.getGlobal(offset->name)->imported()) {
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/tools/fuzzing.h:487]: I'll improve the comment to this:
```
        // It is legal to use an imported global as a segment initializer, but
        // *not* to use a non-imported global. We have made all globals be
        // non-imported earlier (to avoid having to handle arbitrary imports
        // in the fuzzer harness), which means we may refer to a global here
        // that used to be imported, but no longer is, which we must fix up.
```

File: src/tools/fuzzing.h [hunk 17]
```diff
@@ -432,21 +461,83 @@ class TranslateToFuzzReader {
...
+            // TODO: the segments must not overlap...
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/tools/fuzzing.h]: Hmm, you're right, it is valid wasm. It would prevent MemoryPacking from running though, so we may want to avoid it anyhow. I'll add that to the comment.

File: src/tools/fuzzing.h (hunk 6 above)
Comment (kripken) [reply src/tools/fuzzing.h]: Done.

File: src/tools/fuzzing.h [hunk 18]
```diff
@@ -432,21 +461,83 @@ class TranslateToFuzzReader {
...
+    // See above for globals.
+    wasm.table.module = wasm.table.base = Name();
   }
 
-  const Name HANG_LIMIT_GLOBAL = "hangLimit";
+  Name HANG_LIMIT_GLOBAL;
+
+  void prepareHangLimitSupport() {
+    HANG_LIMIT_GLOBAL = Names::getValidGlobalName(wasm, "hangLimit");
+  }
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/tools/fuzzing.h:546]: The prepare step runs first and picks a global name for the hang limit global. We need to do that first so that we know it when we create references to that global, which we do throughout code generation in functions - in both initial functions, and new functions.

> If we add the global before parsing the initial content like now, what happens the initial content contains a global with the same name as the hang limit global?

The initial content is present before we reach this file (it is loaded, then we run this code in this file). So we get a valid global name for the hang limit global that does not collide with the initial content here.

We parse the initial content later to modify it with some fuzzing changes, but no names are added at that time.

File: src/tools/fuzzing.h [hunk 19]
```diff
@@ -606,6 +716,10 @@ class TranslateToFuzzReader {
       std::map<Type, std::vector<Expression*>> exprsByType;
 
       void visitExpression(Expression* curr) {
+        // Some things cannot be moved around easily, give up on them.
+        if (!FindAll<Pop>(curr).list.empty()) {
+          return;
+        }
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/tools/fuzzing.h]: Oh, I thought I removed all those, but I guess I missed this one... removing.

File: src/tools/fuzzing.h [hunk 20]
```diff
@@ -650,9 +764,10 @@ class TranslateToFuzzReader {
         if (parent.oneIn(10)) {
           // Replace it!
           auto& candidates = scanner.exprsByType[curr->type];
-          assert(!candidates.empty()); // this expression itself must be there
-          replaceCurrent(
-            ExpressionManipulator::copy(parent.pick(candidates), wasm));
+          if (!candidates.empty()) {
+            replaceCurrent(
+              ExpressionManipulator::copy(parent.pick(candidates), wasm));
+          }
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/tools/fuzzing.h]: I think this went with the pop change, which I'm removing. (A pop would not be added.) So this can be removed.

File: src/tools/fuzzing.h (hunk 8 above)
Comment (kripken) [reply src/tools/fuzzing.h:908]: Yes, basically this makes it so sometimes we fuzz most functions, and sometimes few (and not always around the same % in a large enough module due to statistics). Not sure it matters much though.

File: src/tools/fuzzing.h (hunk 9 above)
Comment (kripken) [reply src/tools/fuzzing.h:922]: I meant that b returns something that a consumes. could be `(i32.eqz (i32.const 5))`. (but this is just a vague idea...)

File: src/tools/fuzzing.h (hunk 10 above)
Comment (kripken) [reply src/tools/fuzzing.h:930]: It would not help much to check that here. When we generate code ourselves, we can add OOB checks into it as we create it. So it makes sense to check there. But for initial contents, we didn't create them and they don't have OOB checks anyhow, so it's likely they can have OOB which will trap (this is a downside of initial contents, atm). But it's still important to modify them, which helps finds bugs, and doesn't increase the chance of OOB.

You made me realize that if we move to add OOB checks later, we'd want to change this code (as we could then add them even to initial contents). I added a TODO.

File: test/unit/input/hello_world.wat [hunk 21]
```diff
@@ -0,0 +1,11 @@
+(module
```
[the remark this answers is not in the document]
Comment (kripken) [reply test/unit/input/hello_world.wat:1]: heh, it's similar to the existing https://github.com/WebAssembly/binaryen/blob/master/test/hello_world.wat - kind of a minimal wasm. if wasm had strings, it could print hello world...

File: scripts/fuzz_opt.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (aheejin) [reply scripts/fuzz_opt.py]: Printing index sounds good. Do you think the size info is not worth displaying? (I don't have a particular opinion here; if you think it's not that should be fine)

File: src/tools/fuzzing.h [hunk 22]
```diff
@@ -432,21 +463,87 @@ class TranslateToFuzzReader {
   void setupEvents() {
     Index num = upTo(3);
     for (size_t i = 0; i < num; i++) {
-      auto* event =
-        builder.makeEvent(std::string("event$") + std::to_string(i),
-                          WASM_EVENT_ATTRIBUTE_EXCEPTION,
-                          Signature(getControlFlowType(), Type::none));
-      wasm.addEvent(event);
+      auto name = Names::getValidEventName(wasm, "event$");
+      if (!wasm.getEventOrNull(name)) {
```
Comment (aheejin) [inline src/tools/fuzzing.h:467]: We don't need this check, because `getValidEventName` guarantees it

File: src/tools/fuzzing.h (hunk 16 above)
[the remark this answers is not in the document]
Comment (aheejin) [reply src/tools/fuzzing.h:487]: Thanks. But isn't `if (!wasm.getGlobal(offset->name)->imported())` always true? Do we need this check?

File: src/tools/fuzzing.h (hunk 10 above)
Comment (aheejin) [reply src/tools/fuzzing.h:930]: Where do we add OOB checks in the fuzzer now?

**Review by aheejin — APPROVED** [review]

File: scripts/fuzz_opt.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (kripken) [reply scripts/fuzz_opt.py]: Me either. We can add the size later if we want, I guess.

File: src/tools/fuzzing.h (hunk 16 above)
[the remark this answers is not in the document]
Comment (kripken) [reply src/tools/fuzzing.h:487]: We do make all the globals in the initial contents be non-imported. But there may be other globals that the fuzzer adds, that are imported. I think there aren't any atm, but this code would handle that.

I agree it looks confusing atm, I'll add a comment.

File: src/tools/fuzzing.h (hunk 10 above)
Comment (kripken) [reply src/tools/fuzzing.h:930]: In `makePointer` (which emits a pointer for a load/store/etc.) if we don't allow OOB then we mask it to make sure it fits in memory.
